`st.datetimes(allow_imaginary=False)` can raise internal error

`strategies.datetime` may raise such an error when passed `allow_imaginary=False` and tzinfo is None.  

I have the following:

```
dates = datetimes(
    min_value=datetime.datetime(2000, 1,1),
    max_value=datetime.datetime(9000, 12,31),
    allow_imaginary=False,
)
```

The error looks like (pytest style):

```
value = datetime.datetime(2000, 1, 1, 0, 0)

    def datetime_does_not_exist(value):
        # This function tests whether the given datetime can be round-tripped to and
        # from UTC.  It is an exact inverse of (and very similar to) the dateutil method
        # https://dateutil.readthedocs.io/en/stable/tz.html#dateutil.tz.datetime_exists
        try:
            # Does the naive portion of the datetime change when round-tripped to
            # UTC?  If so, or if this overflows, we say that it does not exist.
            roundtrip = value.astimezone(dt.timezone.utc).astimezone(value.tzinfo)
        except OverflowError:
            # Overflows at datetime.min or datetime.max boundary condition.
            # Rejecting these is acceptable, because timezones are close to
            # meaningless before ~1900 and subject to a lot of change by
            # 9999, so it should be a very small fraction of possible values.
            return True
>       assert value.tzinfo is roundtrip.tzinfo, "so only the naive portions are compared"
E       AssertionError: so only the naive portions are compared
E       assert None is datetime.timezone(datetime.timedelta(days=-1, seconds=68400), 'CST')
E        +  where None = datetime.datetime(2000, 1, 1, 0, 0).tzinfo
E        +  and   datetime.timezone(datetime.timedelta(days=-1, seconds=68400), 'CST') = datetime.datetime(2000, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400), 'CST')).tzinfo

.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/datetime.py:75: AssertionError
```
Notice that `value.tzinfo` is None, so `astimezone` uses the local timezone.

--- a/hypothesis-python/RELEASE.rst
+++ b/hypothesis-python/RELEASE.rst
@@ -1,0 +1,5 @@
+RELEASE_TYPE: patch
+
+This patch fixes an internal error in :func:`~hypothesis.strategies.datetimes`
+with ``allow_imaginary=False`` where the ``timezones`` argument can generate
+``tzinfo=None`` (:issue:`2662`).

--- a/hypothesis-python/src/hypothesis/strategies/_internal/datetime.py
+++ b/hypothesis-python/src/hypothesis/strategies/_internal/datetime.py
@@ -59,9 +59,15 @@ def replace_tzinfo(value, timezone):
 
 
 def datetime_does_not_exist(value):
-    # This function tests whether the given datetime can be round-tripped to and
-    # from UTC.  It is an exact inverse of (and very similar to) the dateutil method
-    # https://dateutil.readthedocs.io/en/stable/tz.html#dateutil.tz.datetime_exists
+    """This function tests whether the given datetime can be round-tripped to and
+    from UTC.  It is an exact inverse of (and very similar to) the dateutil method
+    https://dateutil.readthedocs.io/en/stable/tz.html#dateutil.tz.datetime_exists
+    """
+    # Naive datetimes cannot be imaginary, but we need this special case because
+    # chaining .astimezone() ends with *the system local timezone*, not None.
+    # See bug report in https://github.com/HypothesisWorks/hypothesis/issues/2662
+    if value.tzinfo is None:
+        return False
     try:
         # Does the naive portion of the datetime change when round-tripped to
         # UTC?  If so, or if this overflows, we say that it does not exist.

--- a/hypothesis-python/tests/cover/test_datetimes.py
+++ b/hypothesis-python/tests/cover/test_datetimes.py
@@ -147,3 +147,8 @@ def test_can_generate_datetime_with_fold_1():
 @pytest.mark.skipif(PYPY, reason="see comment")
 def test_can_generate_time_with_fold_1():
     find_any(times(), lambda d: d.fold)
+
+
+@given(datetimes(allow_imaginary=False))
+def test_allow_imaginary_is_not_an_error_for_naive_datetimes(d):
+    pass

--- a/pytest.ini
+++ b/pytest.ini
@@ -1,3 +1,5 @@
 [pytest]
 
 addopts=--strict --tb=native -p pytester --runpytest=subprocess --durations=20
+filterwarnings =
+    ignore:The `.example()` method is good for exploring strategies, but should only be used interactively

